File: Xdows.Native.Script/ScriptScan.cpp
#include "pch.h"
#include "ScriptScan.hpp"

#include <string>
#include <vector>
#include <set>
#include <algorithm>
#include <sstream>

#define PCRE2_CODE_UNIT_WIDTH 8
#include <pcre2.h>

namespace
{
    const uint32_t MAX_CONTENT_SIZE = 10485760;
    const uint32_t MAX_SCORE = 1000;

    const std::set<std::wstring>& GetScriptExtensions()
    {
        static const std::set<std::wstring> extensions = {
            L".ps1", L".psm1", L".psd1",
            L".vbs", L".vbe",
            L".js", L".jse",
            L".bat", L".cmd",
            L".py", L".pyw",
            L".sh", L".bash", L".zsh",
            L".pl", L".pm",
            L".rb",
            L".php", L".phtml", L".php3", L".php4", L".php5",
            L".lnk"
        };
        return extensions;
    }

    struct DetectionPattern
    {
        std::wstring Pattern;
        int Score;
        std::wstring Tag;
        pcre2_code* CompiledPattern;
    };

    std::vector<DetectionPattern> g_genericPatterns;
    std::vector<DetectionPattern> g_powershellPatterns;
    std::vector<DetectionPattern> g_batchPatterns;
    bool g_patternsInitialized = false;

    std::wstring ToLower(const std::wstring& str)
    {
        std::wstring result = str;
        std::transform(result.begin(), result.end(), result.begin(), ::towlower);
        return result;
    }

    std::string ToUtf8(const std::wstring& wstr)
    {
        if (wstr.empty()) return std::string();
        
        int size = WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), -1, nullptr, 0, nullptr, nullptr);
        std::string result(size - 1, 0);
        WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), -1, result.data(), size, nullptr, nullptr);
        return result;
    }

    pcre2_code* CompilePattern(const std::wstring& pattern)
    {
        std::string utf8Pattern = ToUtf8(pattern);
        
        int errorNumber = 0;
        PCRE2_SIZE errorOffset = 0;
        
        pcre2_code* re = pcre2_compile(
            reinterpret_cast<PCRE2_SPTR>(utf8Pattern.c_str()),
            PCRE2_ZERO_TERMINATED,
            PCRE2_CASELESS,
            &errorNumber,
            &errorOffset,
            nullptr);
        
        return re;
    }

    void InitializePatterns()
    {
        if (g_patternsInitialized) return;

        struct PatternDef { std::wstring pattern; int score; std::wstring tag; };
        
        std::vector<PatternDef> genericDefs = {
            {L"eval\\(|Invoke-Expression|Execute\\(|exec\\(", 20, L"DynamicExecution"},
            {L"base64|FromBase64String|atob|btoa", 15, L"EncodedContent"},
            {L"(download|wget|curl|invoke-webrequest|fetch\\s*\\()", 20, L"DownloadBehavior"},
            {L"(http|https|ftp)://", 10, L"NetworkActivity"},
            {L"(delete|remove|copy|move|create\\s+file|write\\s+file)", 10, L"FileOperation"},
            {L"(reg\\s+|registry|regedit|reg.exe)", 15, L"RegistryOperation"},
            {L"(start-process|createobject|wscript.shell|shell.application)", 15, L"ProcessOperation"},
            {L"(startup|runonce|autorun|msconfig)", 20, L"PersistenceMechanism"},
            {L"(nyancat|rainbow|memz|trollface)", 30, L"MEMZSignature"},
            {L"(delete\\s+.*system|format\\s+|shutdown|reboot|blue\\s+screen)", 25, L"SystemDestruction"}
        };

        for (const auto& def : genericDefs)
        {
            pcre2_code* compiled = CompilePattern(def.pattern);
            if (compiled)
            {
                g_genericPatterns.push_back({def.pattern, def.score, def.tag, compiled});
            }
        }

        std::vector<PatternDef> powershellDefs = {
            {L"-executionpolicy\\s+bypass", 20, L"BypassExecutionPolicy"},
            {L"-windowstyle\\s+hidden", 15, L"HiddenWindow"},
            {L"(reflection|assembly.load|loadfrom)", 15, L"ReflectionUsage"},
            {L"(add-type|dllimport|getmodulehandle)", 15, L"WinAPIUsage"},
            {L"new-object\\s+-comobject", 10, L"COMObjectUsage"}
        };

        for (const auto& def : powershellDefs)
        {
            pcre2_code* compiled = CompilePattern(def.pattern);
            if (compiled)
            {
                g_powershellPatterns.push_back({def.pattern, def.score, def.tag, compiled});
            }
        }

        std::vector<PatternDef> batchDefs = {
            {L"@echo\\s+off", 5, L"HiddenCommands"},
            {L"powershell\\s+", 10, L"PowerShellInBatch"},
            {L"certutil\\s+", 15, L"CertutilUsage"},
            {L"bitsadmin\\s+", 15, L"BitsadminUsage"},
            {L"(del\\s+[/sfq]|format\\s+|rmdir\\s+[/sq]|shutdown\\s+[/sfr])", 25, L"SystemDestruction"},
            {L"(reg\\s+(add|delete)|regedit)", 20, L"RegistryModification"}
        };

        for (const auto& def : batchDefs)
        {
            pcre2_code* compiled = CompilePattern(def.pattern);
            if (compiled)
            {
                g_batchPatterns.push_back({def.pattern, def.score, def.tag, compiled});
            }
        }

        g_patternsInitialized = true;
    }

    bool MatchPattern(const std::string& content, const DetectionPattern& pattern)
    {
        if (pattern.CompiledPattern == nullptr) return false;

        pcre2_match_data* matchData = pcre2_match_data_create_from_pattern(pattern.CompiledPattern, nullptr);
        if (matchData == nullptr) return false;

        int rc = pcre2_match(
            pattern.CompiledPattern,
            reinterpret_cast<PCRE2_SPTR>(content.c_str()),
            content.length(),
            0,
            0,
            matchData,
            nullptr);

        pcre2_match_data_free(matchData);
        return rc >= 0;
    }

    void ScanWithPatterns(const std::string& content, const std::vector<DetectionPattern>& patterns, int& totalScore, std::set<std::wstring>& tags)
    {
        for (const auto& pattern : patterns)
        {
            if (MatchPattern(content, pattern))
            {
                totalScore += pattern.Score;
                tags.insert(pattern.Tag);
            }
        }
    }

    std::wstring GetFileExtension(const std::wstring& filePath)
    {
        size_t pos = filePath.rfind(L'.');
        if (pos != std::wstring::npos)
        {
            return ToLower(filePath.substr(pos));
        }
        return L"";
    }

    bool IsPowerShellFile(const std::wstring& ext)
    {
        return ext == L".ps1" || ext == L".psm1" || ext == L".psd1";
    }

    bool IsBatchFile(const std::wstring& ext)
    {
        return ext == L".bat" || ext == L".cmd";
    }

    std::string ConvertToUtf8(const uint8_t* content, size_t size)
    {
        if (size >= 3 && content[0] == 0xEF && content[1] == 0xBB && content[2] == 0xBF)
        {
            return std::string(reinterpret_cast<const char*>(content + 3), size - 3);
        }

        if (size >= 2 && content[0] == 0xFF && content[1] == 0xFE)
        {
            size_t charCount = (size - 2) / 2;
            std::wstring wstr(reinterpret_cast<const wchar_t*>(content + 2), charCount);
            return ToUtf8(wstr);
        }

        if (size >= 2 && content[0] == 0xFE && content[1] == 0xFF)
        {
            std::string result;
            for (size_t i = 2; i + 1 < size; i += 2)
            {
                wchar_t wc = (static_cast<wchar_t>(content[i]) << 8) | content[i + 1];
                char utf8Buf[4] = {0};
                int len = WideCharToMultiByte(CP_UTF8, 0, &wc, 1, utf8Buf, 4, nullptr, nullptr);
                result.append(utf8Buf, len);
            }
            return result;
        }

        return std::string(reinterpret_cast<const char*>(content), size);
    }
}

int ScanScriptFile(const wchar_t* filePath, const uint8_t* content, size_t contentSize, ScriptScanResult* result)
{
    if (result == nullptr)
    {
        return 0;
    }

    result->Score = 0;
    result->Tags[0] = L'\0';

    InitializePatterns();

    if (content == nullptr || contentSize == 0)
    {
        return 1;
    }

    size_t scanSize = (std::min)(contentSize, static_cast<size_t>(MAX_CONTENT_SIZE));
    std::string utf8Content = ConvertToUtf8(content, scanSize);

    std::wstring ext;
    if (filePath != nullptr)
    {
        ext = GetFileExtension(filePath);
    }

    int totalScore = 0;
    std::set<std::wstring> tags;

    ScanWithPatterns(utf8Content, g_genericPatterns, totalScore, tags);

    if (IsPowerShellFile(ext))
    {
        ScanWithPatterns(utf8Content, g_powershellPatterns, totalScore, tags);
    }
    else if (IsBatchFile(ext))
    {
        ScanWithPatterns(utf8Content, g_batchPatterns, totalScore, tags);
    }

    result->Score = (std::min)(totalScore, static_cast<int>(MAX_SCORE));

    std::wostringstream tagStream;
    bool firstTag = true;
    for (const auto& tag : tags)
    {
        if (!firstTag)
        {
            tagStream << L" ";
        }
        firstTag = false;
        tagStream << tag;
    }

    std::wstring tagStr = tagStream.str();
    if (tagStr.size() < 512)
    {
        wcscpy_s(result->Tags, 512, tagStr.c_str());
    }

    return 1;
}

int IsScriptFile(const wchar_t* extension)
{
    if (extension == nullptr)
    {
        return 0;
    }

    std::wstring ext = ToLower(extension);
    const auto& extensions = GetScriptExtensions();

    return extensions.find(ext) != extensions.end() ? 1 : 0;
}

File: Xdows-Local/ScriptScan.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Xdows_Local
{
    public static partial class ScriptScan
    {
        private static readonly Regex RxDownload = new(@"(download|wget|curl|invoke-webrequest|fetch\s*\()", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxUrl = new(@"(http|https|ftp)://", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxFileOp = new(@"(delete|remove|copy|move|create\s+file|write\s+file)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxRegistry = new(@"(reg\s+|registry|regedit|reg.exe)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxProcessOp = new(@"(start-process|createobject|wscript.shell|shell.application)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPersistence = new(@"(startup|runonce|autorun|msconfig)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxMemz = new(@"(nyancat|rainbow|memz|trollface)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxDestruction = new(@"(delete\s+.*system|format\s+|shutdown|reboot|blue\s+screen)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPsBypass = new(@"-executionpolicy\s+bypass", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPsHidden = new(@"-windowstyle\s+hidden", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPsReflection = new(@"(reflection|assembly.load|loadfrom)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPsWinApi = new(@"(add-type|dllimport|getmodulehandle)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPsCom = new(@"new-object\s+-comobject", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxVbWscript = new(@"createobject\s*\(\s*""wscript.shell""", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxVbFso = new(@"createobject\s*\(\s*""scripting.filesystemobject""", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxVbShellApp = new(@"createobject\s*\(\s*""shell.application""", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxJsActiveX = new(@"new\s+activexobject", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxJsWscript = new(@"wscript.", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxBatEchoOff = new(@"@echo\s+off", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxBatPs = new(@"powershell\s+", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxBatCertutil = new(@"certutil\s+", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxBatBitsadmin = new(@"bitsadmin\s+", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxBatDestruction = new(@"(del\s+[/sfq]|format\s+|rmdir\s+[/sq]|shutdown\s+[/sfr])", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxBatReg = new(@"(reg\s+(add|delete)|regedit)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPyOs = new(@"os.system\s*\(", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPySubprocess = new(@"subprocess.", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPyUrllib = new(@"urllib.", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxPyRequests = new(@"requests.", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxShDownload = new(@"(wget|curl)\s+", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        private static readonly Regex RxShChmod = new(@"chmod\s+", RegexOptions.IgnoreCase | RegexOptions.Compiled);
        public static (Int32 score, String extra) ScanScriptFile(String filePath, Byte[] fileContent)
        {
            return ScanScriptFileManaged(filePath, fileContent);
        }

        private static (Int32 score, String extra) ScanScriptFileManaged(String filePath, Byte[] fileContent)
        {
            Int32 score = 0;
            List<String> extra = [];
            String fileExtension = GetExtString(filePath);

            if (IsSuspiciousBat(fileContent))
            {
                score += 10;
                extra.Add("CamouflageBat");
            }

            if (fileExtension == ".lnk")
            {
                (Int32 score, String extra) lnkResult = CheckShortcutFile(filePath, fileContent);
                score += lnkResult.score;
                if (!String.IsNullOrEmpty(lnkResult.extra))
                    extra.Add(lnkResult.extra);
            }
            else if (IsScriptFile(fileExtension))
            {
                (Int32 score, String extra) scriptResult = CheckScriptFile(fileExtension, fileContent);
                score += scriptResult.score;
                if (!String.IsNullOrEmpty(scriptResult.extra))
                    extra.Add(scriptResult.extra);
            }

            return (score, String.Join(" ", extra));
        }

        private static unsafe String GetExtString(String path)
        {
            if (String.IsNullOrEmpty(path)) return String.Empty;

            fixed (Char* p = path)
            {
                Char* dot = null, slash = p;
                for (Char* c = p + path.Length - 1; c >= p; c--)
                {
                    if (*c == '.') { dot = c; break; }
                    if (*c is '\\' or '/') slash = c;
                }
                if (dot == null || dot < slash) return String.Empty;

                Int32 len = (Int32)(p + path.Length - dot);
                Span<Char> buf = stackalloc Char[len];
                ReadOnlySpan<Char> src = new(dot, len);
                src.ToLowerInvariant(buf);
                return buf.ToString();
            }
        }

        private static Boolean IsSuspiciousBat(Byte[] fileContent)
        {
            if (fileContent.Length == 0) return false;
            ReadOnlySpan<Byte> data = fileContent.AsSpan();

            if (data.IndexOf("program cannot be run in"u8) >= 0) return true;
            if (data.IndexOf("LoadLibraryA"u8) >= 0) return true;
            if (data.IndexOf("Win32"u8) >= 0) return true;
            if (data.IndexOf("kernel32.dll"u8) >= 0) return true;
            if (data.IndexOf("ntdll.dll"u8) >= 0) return true;
            if (data.IndexOf("GetProcAddress"u8) >= 0) return true;
            if (data.IndexOf(@"C:\windows\"u8) >= 0) return true;
            if (data.IndexOf("*.exe"u8) >= 0) return true;
            if (data.IndexOf("Shutdown"u8) >= 0) return true;

            return false;
        }

        private static (Int32 score, String extra) CheckShortcutFile(String filePath, Byte[] fileContent)
        {
            Int32 score = 0;
            List<String> extra = [];

            try
            {
                if (fileContent.Length > 1024 * 10)
                {
                    score += 10;
                    extra.Add("LargeShortcut");
                }

                String content = Encoding.ASCII.GetString(fileContent);

                if (content.Contains(".exe") &&
                    (!content.Contains("System32", StringComparison.OrdinalIgnoreCase) &&
                     !content.Contains("Program Files", StringComparison.OrdinalIgnoreCase)))
                {
                    score += 15;
                    extra.Add("SuspiciousTarget");
                }

                if (content.Contains("powershell", StringComparison.OrdinalIgnoreCase) ||
                    content.Contains("cmd.exe", StringComparison.OrdinalIgnoreCase) ||
                    content.Contains("wscript.exe", StringComparison.OrdinalIgnoreCase) ||
                    content.Contains("cscript.exe", StringComparison.OrdinalIgnoreCase))
                {
                    score += 20;
                    extra.Add("ScriptInShortcut");
                }

                if (content.Contains("-windowstyle hidden", StringComparison.OrdinalIgnoreCase) ||
                    content.Contains("-w hidden", StringComparison.OrdinalIgnoreCase))
                {
                    score += 15;
                    extra.Add("HiddenExecution");
                }

                if (content.Contains("base64", StringComparison.OrdinalIgnoreCase) ||
                    content.Contains("FromBase64String", StringComparison.OrdinalIgnoreCase))
                {
                    score += 25;
                    extra.Add("EncodedContent");
                }

                if (content.Contains("download", StringComparison.OrdinalIgnoreCase) ||
                    content.Contains("invoke-webrequest", StringComparison.OrdinalIgnoreCase) ||
                    content.Contains("wget", StringComparison.OrdinalIgnoreCase) ||
                    content.Contains("curl", StringComparison.OrdinalIgnoreCase))
                {
                    score += 20;
                    extra.Add("DownloadBehavior");
                }
            }
            catch
            {
                score += 10;
                extra.Add("CorruptedShortcut");
            }

            return (score, String.Join(" ", extra));
        }

        private static (Int32 score, String extra) CheckScriptFile(String extension, Byte[] fileContent)
        {
            Int32 score = 0;
            List<String> extra = [];

            try
            {
                String content = Encoding.UTF8.GetString(fileContent);

                score += CheckGenericScript(content, extra);

                score += extension switch
                {
                    ".ps1" or ".psm1" or ".psd1" => CheckPowerShellScript(content, extra),
                    ".vbs" or ".vbe" => CheckVBScript(content, extra),
                    ".js" or ".jse" => CheckJavaScript(content, extra),
                    ".bat" or ".cmd" => CheckBatchScript(content, extra),
                    ".py" or ".pyw" => CheckPythonScript(content, extra),
                    ".sh" => CheckShellScript(content, extra),
                    _ => 0
                };
            }
            catch
            {
                score += 10;
                extra.Add("CorruptedScript");
            }

            return (score, String.Join(" ", extra));
        }

        private static Int32 CheckGenericScript(String content, List<String> extra)
        {
            Int32 score = 0;

            if (content.Contains("eval(") || content.Contains("Invoke-Expression") ||
                content.Contains("Execute(") || content.Contains("exec("))
            {
                score += 20;
                extra.Add("DynamicExecution");
            }

            if (content.Contains("base64") || content.Contains("FromBase64String") ||
                content.Contains("atob") || content.Contains("btoa"))
            {
                score += 15;
                extra.Add("EncodedContent");
            }

            if (RxDownload.IsMatch(content))
            {
                score += 20;
                extra.Add("DownloadBehavior");
            }

            if (RxUrl.IsMatch(content))
            {
                score += 10;
                extra.Add("NetworkActivity");
            }

            if (RxFileOp.IsMatch(content))
            {
                score += 10;
                extra.Add("FileOperation");
            }

            if (RxRegistry.IsMatch(content))
            {
                score += 15;
                extra.Add("RegistryOperation");
            }

            if (RxProcessOp.IsMatch(content))
            {
                score += 15;
                extra.Add("ProcessOperation");
            }

            if (RxPersistence.IsMatch(content))
            {
                score += 20;
                extra.Add("PersistenceMechanism");
            }

            if (RxMemz.IsMatch(content))
            {
                score += 30;
                extra.Add("MEMZSignature");
            }

            if (RxDestruction.IsMatch(content))
            {
                score += 25;
                extra.Add("SystemDestruction");
            }

            return score;
        }

        private static Int32 CheckPowerShellScript(String content, List<String> extra)
        {
            Int32 score = 0;

            if (RxPsBypass.IsMatch(content))
            {
                score += 20;
                extra.Add("BypassExecutionPolicy");
            }

            if (RxPsHidden.IsMatch(content))
            {
                score += 15;
                extra.Add("HiddenWindow");
            }

            if (RxPsReflection.IsMatch(content))
            {
                score += 15;
                extra.Add("ReflectionUsage");
            }

            if (RxPsWinApi.IsMatch(content))
            {
                score += 15;
                extra.Add("WinAPIUsage");
            }

            if (RxPsCom.IsMatch(content))
            {
                score += 10;
                extra.Add("COMObjectUsage");
            }

            return score;
        }

        private static Int32 CheckVBScript(String content, List<String> extra)
        {
            Int32 score = 0;

            if (RxVbWscript.IsMatch(content))
            {
                score += 15;
                extra.Add("WScriptShellUsage");
            }

            if (RxVbFso.IsMatch(content))
            {
                score += 10;
                extra.Add("FileSystemObjectUsage");
            }

            if (RxVbShellApp.IsMatch(content))
            {
                score += 15;
                extra.Add("ShellApplicationUsage");
            }

            return score;
        }

        private static Int32 CheckJavaScript(String content, List<String> extra)
        {
            Int32 score = 0;

            if (RxJsActiveX.IsMatch(content))
            {
                score += 15;
                extra.Add("ActiveXObjectUsage");
            }

            if (RxJsWscript.IsMatch(content))
            {
                score += 15;
                extra.Add("WScriptUsage");
            }

            return score;
        }

        private static Int32 CheckBatchScript(String content, List<String> extra)
        {
            Int32 score = 0;

            if (RxBatEchoOff.IsMatch(content))
            {
                score += 5;
                extra.Add("HiddenCommands");
            }

            if (RxBatPs.IsMatch(content))
            {
                score += 10;
                extra.Add("PowerShellInBatch");
            }

            if (RxBatCertutil.IsMatch(content))
            {
                score += 15;
                extra.Add("CertutilUsage");
            }

            if (RxBatBitsadmin.IsMatch(content))
            {
                score += 15;
                extra.Add("BitsadminUsage");
            }

            if (RxBatDestruction.IsMatch(content))
            {
                score += 25;
                extra.Add("SystemDestruction");
            }

            if (RxBatReg.IsMatch(content))
            {
                score += 20;
                extra.Add("RegistryModification");
            }

            return score;
        }

        private static Int32 CheckPythonScript(String content, List<String> extra)
        {
            Int32 score = 0;

            if (RxPyOs.IsMatch(content))
            {
                score += 10;
                extra.Add("OSSystemUsage");
            }

            if (RxPySubprocess.IsMatch(content))
            {
                score += 10;
                extra.Add("SubprocessUsage");
            }

            if (RxPyUrllib.IsMatch(content))
            {
                score += 10;
                extra.Add("UrllibUsage");
            }

            if (RxPyRequests.IsMatch(content))
            {
                score += 10;
                extra.Add("RequestsUsage");
            }

            return score;
        }

        private static Int32 CheckShellScript(String content, List<String> extra)
        {
            Int32 score = 0;

            if (RxShDownload.IsMatch(content))
            {
                score += 10;
                extra.Add("DownloadTool");
            }

            if (RxShChmod.IsMatch(content))
            {
                score += 5;
                extra.Add("ChmodUsage");
            }

            return score;
        }

        private static readonly HashSet<String> _scriptExtensions = new(StringComparer.OrdinalIgnoreCase)
        {
            ".ps1", ".psm1", ".psd1",
            ".vbs", ".vbe",
            ".js", ".jse",
            ".bat", ".cmd",
            ".py", ".pyw",
            ".sh", ".bash", ".zsh",
            ".pl", ".pm",
            ".rb",
            ".php", ".phtml", ".php3", ".php4", ".php5"
        };

        private static Boolean IsScriptFile(String extension)
        {
            return _scriptExtensions.Contains(extension);
        }
    }
}
